### PR TITLE
autogold for search test

### PR DESCRIPTION
**Pull request by rvantonder**
Fewer lines: `+146 −171`

Before:

```
rvt:query/ (main*) $ go test                                                                                                                                                             [14:05:13]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	0.139s
rvt:query/ (main*) $ go test                                                                                                                                                             [14:05:14]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	0.139s
rvt:query/ (main*) $ go test                                                                                                                                                             [14:05:15]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	0.140s
rvt:query/ (main*) $ go test                                                                                                                                                             [14:05:17]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	0.138s
rvt:query/ (main*) $ go test                                                                                                                                                             [14:05:20]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	0.143s
rvt:query/ (main*) $
```

After:

```
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	3.183s
rvt:query/ (rvt/autogold-search-tests*) $ go test                                                                                                                                        [14:06:30]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	3.109s
rvt:query/ (rvt/autogold-search-tests*) $ go test                                                                                                                                        [14:06:36]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	3.488s
rvt:query/ (rvt/autogold-search-tests*) $ go test                                                                                                                                        [14:06:41]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	3.125s
rvt:query/ (rvt/autogold-search-tests*) $ go test                                                                                                                                        [14:06:46]
PASS
ok  	github.com/sourcegraph/sourcegraph/internal/search/query	3.465s
rvt:query/ (rvt/autogold-search-tests*) $ 
```

### CODE REVIEW

**Comment by slimsag** [comment]
Interesting slowness, I suspect that [stringifying the input Go values to ASTs/strings and diffing](https://github.com/hexops/autogold/blob/main/want.go#L93-L96) must be quite expensive for some reason here. I suspect there is some very low-hanging fruit here I can address, I haven't tried to optimize anything here really - I'll try poking at this a bit after work to see how much faster I can make it.

File: internal/search/query/parser_test.go [hunk 1]
```diff
@@ -28,183 +29,157 @@ func heuristicLabels(nodes []Node) string {
...
+		WantLabels: labels(10),
+		WantRange:  "{\"start\":{\"line\":0,\"column\":0},\"end\":{\"line\":0,\"column\":8}}",
+	}).Equal(t, test(`"quoted"`))
+
+	autogold.Want("911d89f74e70", value{
+		Want: "{\"value\":\"'\",\"negated\":false}", WantLabels: labels(10),
+		WantRange: "{\"start\":{\"line\":0,\"column\":0},\"end\":{\"line\":0,\"column\":4}}",
+	}).Equal(t, test(`'\''`))
+
+	autogold.Want("b58bf1c4c1dc", value{
```
Comment (slimsag) [inline internal/search/query/parser_test.go:160]: Is `b58bf1c4c1dc` a generated ID or something related to this test? I'm guessing generated?

I have some plans to remove the need to specify a name parameter here at all, which I can maybe accelerate if it's particularly painful for your usage here

**Comment by slimsag** [comment]
Cool to see you trying to use this!

File: internal/search/query/parser_test.go (hunk 1 above)
Comment (rvantonder) [reply internal/search/query/parser_test.go:160]: Yeah I just generated it to capture all the tests quickly, not all of them had names. I will give these sensible names later.

**Comment by rvantonder** [comment]
FYI @slimsag in [this commit](https://github.com/sourcegraph/sourcegraph/commit/274abc5c0c13abae7c6ede8399fbe3e36895a879) I experimented with making the compared value just a string (go values are marshaled to JSON) and it results in the same timing, no speed up: (about 3 seconds). But string comparisons can't be that expensive, so I think there's something else afoot with the slowdown. Honestly, it's probably a good thing it's still slow with strings, because that means it's probably not an inherently slow compare operation, just some inefficiency around the comparisons.

**Comment by slimsag** [comment]
Oh slick! That means whatever this is will be really easy to fix. I bet it's calculating the diff when a simple string equality check would work. I'll try that out and report back in like an hour once I'm off work / finish up my current stuffs

**Comment by rvantonder** [comment]
This [line](https://sourcegraph.com/github.com/hexops/autogold/-/blob/want.go#L84) is the culprit:

```go
                        pkgs, err := packages.Load(&packages.Config{Mode: packages.NeedName}, filepath.Dir(testPath))
```

It takes 132ms on my machine, per test. Since this info is needed only per file, should just find a way to frontload this per-file processing. Then we should be good (the other operations in this function are negligible).

**Comment by slimsag** [comment]
I think https://github.com/hexops/autogold/commit/1457509508e1e80b2a1ed4df174d71449b947b88 will fix the issue, you can `go get -u github.com/hexops/autogold@1457509` to try it out

**Comment by rvantonder** [comment]
Woop, this runs in 300ms now. I will update all the tests tomorrow and put up the PR for reals.

**Comment by rvantonder** [comment]
doing in a fresh pr, lots of comments here :3

**Pull request closed without merging**